assembly/__tests__/access_control.spec.ts

```diff
@@ -0,0 +1,255 @@
+    }).toThrow('User does not have admin or user permission');
+  });
+
+  test('should add permissions to user', () => {
+    resetStorage();
+    const accessControl = new AccessControl<u8>(1);
+    const ADMIN = accessControl.newPermission('admin');
+    const USER = accessControl.newPermission('user');
+    const GUEST = accessControl.newPermission('guest');
+    const userAddress = new Address(
+      'AU12UBnqTHDQALpocVBnkPNy7y5CndUJQTLutaVDDFgMJcq5kQiKq',
+    );
+    accessControl.grantPermission(USER, userAddress);
+    accessControl.grantPermission(GUEST, userAddress);
+
+    expect(accessControl.hasPermission(USER, userAddress)).toBeTruthy(
+      'User should have user permission',
+    );
+    expect(accessControl.hasPermission(GUEST, userAddress)).toBeTruthy(
+      'User should have guest permission',
+    );
+    expect(accessControl.hasPermission(ADMIN, userAddress)).toBeFalsy(
+      'User should not have admin permission',
+    );
+  });
+
+  test('should remove permissions from user', () => {
+    resetStorage();
+    const accessControl = new AccessControl<u8>(1);
+    const ADMIN = accessControl.newPermission('admin');
+    const USER = accessControl.newPermission('user');
+    const GUEST = accessControl.newPermission('guest');
+    const userAddress = new Address(
+      'AU12UBnqTHDQALpocVBnkPNy7y5CndUJQTLutaVDDFgMJcq5kQiKq',
+    );
+    accessControl.grantPermission(USER, userAddress);
+    accessControl.grantPermission(GUEST, userAddress);
+    accessControl.grantPermission(ADMIN, userAddress);
+
+    expect(accessControl.hasPermission(USER, userAddress)).toBeTruthy(
+      'User should have user permission',
+    );
+    expect(accessControl.hasPermission(GUEST, userAddress)).toBeTruthy(
+      'User should have guest permission',
+    );
+    expect(accessControl.hasPermission(ADMIN, userAddress)).toBeTruthy(
+      'User should have admin permission',
+    );
+
+    accessControl.revokePermission(USER, userAddress);
+
+    expect(accessControl.hasPermission(USER, userAddress)).toBeFalsy(
+      'User should not have user permission',
+    );
+    expect(accessControl.hasPermission(GUEST, userAddress)).toBeTruthy(
+      'User should have guest permission',
+    );
+    expect(accessControl.hasPermission(ADMIN, userAddress)).toBeTruthy(
+      'User should have admin permission',
+    );
+  });
+
+  test('should return proper throw message', () => {
+    resetStorage();
+    const accessControl = new AccessControl<u8>(1);
+    const ADMIN = accessControl.newPermission('admin');
+    const USER = accessControl.newPermission('user');
+    const GUEST = accessControl.newPermission('guest');
+    const userAddress = new Address(
+      'AU12UBnqTHDQALpocVBnkPNy7y5CndUJQTLutaVDDFgMJcq5kQiKq',
+    );
+
+    // accessControl.mustHavePermission(ADMIN, userAddress);
+    // accessControl.mustHavePermission(USER, userAddress);
+    // accessControl.mustHavePermission(GUEST, userAddress);
+  });
```

Comment:
Is this test useful ?

Reply:
Yes, it let me manually check that the assert message is the correct one. To be remove once this test would be performed by an integration test.